jinahub/indexers/searcher/compound/FaissPostgresSearcher/faisspsql.py

```diff
@@ -2,51 +2,122 @@
-except:
-    from jina_executors.indexers.searcher.FaissSearcher.faiss_searcher import FaissSearcher
+except:  # noqa: E722
+    from jina_executors.indexers.searcher.FaissSearcher.faiss_searcher import (
+        FaissSearcher,
+    )
 
 try:
-    from jinahub.indexers.storage.PostgreSQLStorage import (
-        PostgreSQLStorage,
-    )
-except:
-    from jina_executors.indexers.storage.PostgreSQLStorage import (
-        PostgreSQLStorage,
-    )
+    from jinahub.indexers.storage.PostgreSQLStorage import PostgreSQLStorage
+except:  # noqa: E722
+    from jina_executors.indexers.storage.PostgreSQLStorage import PostgreSQLStorage
 
 
 class FaissPostgresSearcher(Executor):
-    """A Compound Indexer made up of a FaissSearcher (for vectors) and a Postgres Indexer"""
+    """A Compound Indexer made up of a FaissSearcher (for vectors) and a Postgres
+    Indexer
+
+    :param dump_path: a path to a dump folder containing
+    the dump data obtained by calling jina_commons.dump_docs
+    :param use_dump_func: whether to use the dump
+     function of PostgreSQLStorage, when dump_path is not provided
+    """
 
     def __init__(
         self,
-        dump_path=None,
+        dump_path: Optional[str] = None,
+        use_dump_func: bool = False,
+        total_shards: int = None,
         **kwargs,
     ):
         super().__init__(**kwargs)
-        # when constructed from rolling update the dump_path is passed via a runtime_arg
-        dump_path = dump_path or kwargs.get('runtime_args').get('dump_path')
         self.logger = get_logger(self)
+
+        self.total_shards = total_shards
+        if self.total_shards is None:
+            self.logger.warning(
+                'total_shards is None, rolling update '
+                'via PSQL import will not be possible.'
+            )
+
+        # when constructed from rolling update the dump_path is passed via a
+        # runtime_arg
+        dump_path = dump_path or kwargs.get('runtime_args').get('dump_path')
+        use_dump_func = use_dump_func or kwargs.get('runtime_args').get('use_dump_func')
+
         self._kv_indexer = None
         self._vec_indexer = None
-        if dump_path:
-            self._vec_indexer = FaissSearcher(dump_path=dump_path, **kwargs)
+        self._init_kwargs = kwargs
+
+        if dump_path is None and use_dump_func is None:
+            name = getattr(self.metas, 'name', self.__class__.__name__)
+            self.logger.warning(
+                f'No "dump_path" or "use_dump_func" provided '
+                f'for {name}. Use .rolling_update() to re-initialize...'
+            )
+
+        if use_dump_func:
             self._kv_indexer = PostgreSQLStorage(**kwargs)
+            dump_func = self._kv_indexer.get_snapshot
+            self._vec_indexer = FaissSearcher(dump_func=dump_func, **kwargs)
         else:
+            self._kv_indexer = PostgreSQLStorage(**kwargs)
+            self._vec_indexer = FaissSearcher(dump_path=dump_path, **kwargs)
```

Comment:
`self._vec_indexer, self._kv_indexer = self._init_indexers(...)` will reduce complexity here.

Reply:
We need to get the `get_snapshot` of `kv_indexer` to pass to the faiss. So this wouldn't work, right?

Reply:
I just meant to put all the code from line 60-73 into another function, which return the `kv_indexer` and `vec_indexer`

Reply:
I refactored this a bit. Check again